streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java

```diff
@@ -0,0 +1,478 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.assignment.assignors;
+
+import static java.util.Collections.unmodifiableMap;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.assignment.ApplicationState;
+import org.apache.kafka.streams.processor.assignment.KafkaStreamsAssignment;
+import org.apache.kafka.streams.processor.assignment.KafkaStreamsAssignment.AssignedTask;
+import org.apache.kafka.streams.processor.assignment.KafkaStreamsState;
+import org.apache.kafka.streams.processor.assignment.ProcessId;
+import org.apache.kafka.streams.processor.assignment.TaskAssignmentUtils;
+import org.apache.kafka.streams.processor.assignment.TaskAssignor;
+import org.apache.kafka.streams.processor.assignment.TaskInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class StickyTaskAssignor implements TaskAssignor {
+    private static final Logger LOG = LoggerFactory.getLogger(StickyTaskAssignor.class);
+
+    private Map<String, ?> configs = new HashMap<>();
+    private final boolean mustPreserveActiveTaskAssignment;
+
+    public StickyTaskAssignor() {
+        this(false);
+    }
+
+    public StickyTaskAssignor(final boolean mustPreserveActiveTaskAssignment) {
+        this.mustPreserveActiveTaskAssignment = mustPreserveActiveTaskAssignment;
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        // TODO: The application state already has the assignment configs object, why should this
+        //       assignor be configurable?
+        this.configs = configs;
+    }
+
+    @Override
+    public TaskAssignment assign(final ApplicationState applicationState) {
+        final int taskCount = applicationState.allTasks().size();
+        if (taskCount == 0) {
```

Comment:
technically this would be a bug, that should be caught prior to ever invoking the assignor, so I don't think we should/need to check this

Also: I would say that if we did want to return an empty assignment, it should be a real TaskAssignment with a KafkaStreamsAssignment for each KafkaStreamsState that appears in the ApplicationState, but with those just being empty KafkaStreamsAssignment objects (ie no actual tasks assigned).

I actually think it should be considered an error if the returned `TaskAssignment` does not contain a KafkaStreamsAssignment for each KafkaStreamsState in the input. When we get around to implementing the assignment validation and `#onAssignmentComputed` callback with the error codes, we should probably have a dedicated error code for this case (if we don't already, don't remember exactly which error codes we had in the original KIP)

Doesn't matter for now since imo we should remove this check altogether, but good to keep in mind

Reply:
(I filed a ticket for this already btw)